### PR TITLE
feat: added proof of work

### LINKED ISSUE #20: Add support for proof of work
Add required functionality for calculating proof of work given block hash,  transaction ID, difficulty, and hash function name.

### PR DESCRIPTION
**Pull request by ze97286**
Closes #20

### CODE REVIEW

File: libs/crypto/proof_of_work.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		} else {
+			nonce++
+		}
+	}
+
+	return nonce, h[:], nil
+}
+
+// Verify checks that the hash with the given nonce results in the target difficulty.
+func Verify(blockHash string, txID string, nonce int, hashFuncion string, difficulty int) bool {
```
Comment (wwestgarth) [inline libs/crypto/proof_of_work.go]: Should `difficulty` be a `uint`? Feel like we need some checks on `difficulty` too, it panics if you pass in `257`.

File: libs/crypto/proof_of_work.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+func hash(data []byte, hashFunction string) ([]byte, error) {
+	if hashFunction == Sha3 {
+		return Hash(data), nil
+	}
+	return nil, errors.New("unknown hash function")
+}
+
+func IntToHex(num int64) []byte {
```
Comment (jeremyletang) [inline libs/crypto/proof_of_work.go]: Is this really IntToHex, it seems like we just return the bytes of the integer?
IntToHex would be more like this then: `fmt.Sprintf("%x", num)`

File: libs/crypto/proof_of_work.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+const (
+	Sha3     = "sha3_24_rounds"
+	maxNonce = math.MaxInt64
+)
+
+var prefix = []byte("Vega_SPAM_PoW")
+
+// PoW calculates proof of work given block hash, transaction hash, target difficulty and a hash function.
+// returns the nonce, the hash and th error if any.
+func PoW(blockHash string, txID string, difficulty int, hashFunction string) (int, []byte, error) {
```
Comment (jeremyletang) [inline libs/crypto/proof_of_work.go]: Can we have a negative difficulty?

File: libs/crypto/proof_of_work.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	var hashInt big.Int
+	var h []byte
+	var err error
+	nonce := 0
```
Comment (jeremyletang) [inline libs/crypto/proof_of_work.go]: Couldn't / shouldn't we use only uint(64) here?

**Review by wwestgarth — DISMISSED** [review]

File: libs/crypto/proof_of_work.go (hunk 2 above)
Comment (ze97286) [reply libs/crypto/proof_of_work.go]: it's not the same.

File: libs/crypto/proof_of_work.go (hunk 3 above)
Comment (ze97286) [reply libs/crypto/proof_of_work.go]: it's uint now.

File: libs/crypto/proof_of_work.go (hunk 4 above)
Comment (ze97286) [reply libs/crypto/proof_of_work.go]: it is.

**Review by wwestgarth — APPROVED** [review]

**Pull request merged**